### PR TITLE
Add CAVC Power of Attorney admin action

### PR DESCRIPTION
**Pull request by hschallhorn**
Bumps [CASEFLOW-107]

### Description
Allows cavc users to add poa admin action.

### Acceptance Criteria
- [x] The following task actions are added to SendCavcRemandProcessedLetterTask
  - [x]  "Assign POA Clarification task to CAVC team" - Creates a CavcPoaClarificationTask assigned to the CavcLitigationSupportTeam as a child of the SendCavcRemandProcessedLetterTask
- [x] Provide Instructions and context for this action appears on the submit admin actions page
- [x] Users have standard task actions on the poa clarification task (assign, reassign, complete, cancel)

### Testing Plan
1. Log in as the CAVC team admin and assign any open `SendCavcRemandProcessedLetterTask` to yourself
1. Click the task action dropdown and confirm new poa clarification action
1. Confirm you can add instructions to this action
1. Confirm clicking submit keeps you on the case details page and creates a poa clarification task assigned to the cavc team is created as a child of the send letter task
1. Confirm you no longer have actions available to reassign or complete the send letter task
```ruby
Appeal 934 (evidence_submission)                       ID   STATUS      ASGN_BY                ASGN_TO                UPDATED_AT
└── RootTask                                           2841 on_hold                            Bva                    2020-11-23 16:07:47 UTC
    └── DistributionTask                               2842 on_hold                            Bva                    2020-11-23 16:07:47 UTC
        └── CavcTask                                   2843 on_hold                            Bva                    2020-11-23 16:07:47 UTC
            └── SendCavcRemandProcessedLetterTask      2844 on_hold                            CavcLitigationSupport  2020-11-23 17:01:12 UTC
                └── SendCavcRemandProcessedLetterTask  3500 on_hold     CAVC_LIT_SUPPORT_ADMIN CAVC_LIT_SUPPORT_ADMIN 2020-11-23 17:14:51 UTC
                    └── CavcPoaClarificationTask       3501 assigned    CAVC_LIT_SUPPORT_ADMIN CavcLitigationSupport  2020-11-23 17:15:17 UTC
```

### User Facing Changes

 Workflow Step|UI
 ---|---
New task action|Screen Shot 2020-11-23 at 12 01 19 PM
Modal|Screen Shot 2020-11-23 at 12 13 48 PM
Modal validation|Screen Shot 2020-11-23 at 12 14 02 PM
Success & updated task snapshot|Screen Shot 2020-11-23 at 12 14 55 PM
POA clarification admin actions|Screen Shot 2020-11-23 at 1 37 52 PM
POA clarification assignee actions|Screen Shot 2020-11-23 at 1 38 06 PM

### CODE REVIEW

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -17,6 +17,7 @@ class TasksController < ApplicationController
     BlockedSpecialCaseMovementTask: BlockedSpecialCaseMovementTask,
     ChangeHearingDispositionTask: ChangeHearingDispositionTask,
     ColocatedTask: ColocatedTask,
+    CavcPoaClarificationTask: CavcPoaClarificationTask,
```
Comment (hschallhorn) [inline app/controllers/tasks_controller.rb:20]: allows users to create this task type

File: app/models/tasks/cavc_poa_clarification_task.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+# letter to the veteran
+class CavcPoaClarificationTask < Task
+  def self.label
+    COPY::CAVC_POA_TASK_LABEL
+  end
+
+  def available_actions(user)
+    super(user) - [Constants.TASK_ACTIONS.ASSIGN_TO_TEAM.to_h]
+  end
+end
```
Comment (hschallhorn) [inline app/models/tasks/cavc_poa_clarification_task.rb:13]: New task specifically for cavc POA. standard actions other than assign to team https://dsva.slack.com/archives/CJL810329/p1606151814359500

File: app/repositories/task_action_repository.rb [hunk 3]
```diff
@@ -229,6 +229,15 @@ def assign_ihp_to_colocated_blocking_distribution_data(task, _user = nil)
       cavc_add_blocking_admin_action_data(task, Colocated.singleton, IhpColocatedTask)
     end
 
+    def assign_poa_to_cavc_blocking_cavc_data(task, _user = nil)
+      org = CavcLitigationSupport.singleton
+      task_type = CavcPoaClarificationTask
+      cavc_add_blocking_admin_action_data(task, org, task_type).merge(
+        modal_body: format(COPY::CAVC_SEND_TO_TEAM_BLOCKING_CAVC_DETAIL, task_type.label, org.name),
+        parent_id: task.id
```
Comment (hschallhorn) [inline app/repositories/task_action_repository.rb:237]: Should create the task as a child for the send letter task

File: app/models/tasks/cavc_poa_clarification_task.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+# Task for the CAVC Litigation Support team to clarify the Power of Attorney on a remand before sending the 90 day
+# letter to the veteran
```
Comment (yoomlam) [inline app/models/tasks/cavc_poa_clarification_task.rb:4]:
```suggestion
# Task for the CAVC Litigation Support team to clarify the Power of Attorney on a remand before sending the 90 day
# letter to the veteran
# Expected parent: SendCavcRemandProcessedLetterTask
```

File: app/models/tasks/cavc_poa_clarification_task.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+class CavcPoaClarificationTask < Task
```
Comment (yoomlam) [inline app/models/tasks/cavc_poa_clarification_task.rb:5]: Add validation for data integrity
```suggestion
class CavcPoaClarificationTask < Task
  validates :parent, presence: true, parentTask: { task_type: SendCavcRemandProcessedLetterTask }, on: :create
```

File: client/COPY.json [hunk 6]
```diff
@@ -374,6 +374,9 @@
   "SEND_CAVC_REMAND_PROCESSED_LETTER_TASK_LABEL": "Send CAVC-Remand-Processed Letter Task",
 
   "CAVC_SEND_TO_TEAM_NON_BLOCKING_DETAIL": "Create a %s assigned to the %s team. This will not block you from completing this task or sending the 90-day letter to the veteran.",
+  "CAVC_SEND_TO_TEAM_BLOCKING_CAVC_DETAIL": "Create a %s task assigned to the %s team. This will block you from sending the 90-day letter to the veteran until the new task is completed.",
+
+  "CAVC_POA_TASK_LABEL": "Verify Power of Attorney",
```
Comment (yoomlam) [inline client/COPY.json:379]: To be consistent with other verbiage, e.g., `CLARIFY_POA_BLOCKING_CAVC` and "Assign POA Clarification task to CAVC team"
```suggestion
  "CAVC_POA_TASK_LABEL": "Clarify Power of Attorney",
```

File: spec/feature/queue/cavc_task_queue_spec.rb [hunk 7]
```diff
@@ -69,6 +69,29 @@
...
+      expect(send_task_row.find(".taskActionsContainerStyling").all("*", wait: false).length).to be 0
+
+      # Complete the task to unblock
+      click_dropdown(text: Constants.TASK_ACTIONS.MARK_COMPLETE.label)
+      fill_in "completeTaskInstructions", with: "POA verified"
+      click_on COPY::MARK_TASK_COMPLETE_BUTTON
+      visit "queue/appeals/#{send_task.appeal.external_id}"
+      send_task_row = page.find("#currently-active-tasks").find_all("tr")[-2]
+      expect(send_task_row.find(".taskActionsContainerStyling").all("*", wait: false).length).to be > 0
+
```
Comment (yoomlam) [inline spec/feature/queue/cavc_task_queue_spec.rb:94]: Why are the other admin tasks not under the send letter task? The user story suggests these admin tasks are required before sending the letter.
```ruby
                                                      ┌────────────────┐
Appeal 1 (evidence_submission) ────────────────────── │ ID │ STATUS    │
└── RootTask                                          │ 1  │ on_hold   │
    └── DistributionTask                              │ 2  │ on_hold   │
        ├── CavcTask                                  │ 3  │ on_hold   │
        │   └── SendCavcRemandProcessedLetterTask     │ 4  │ on_hold   │
        │       ├── SendCavcRemandProcessedLetterTask │ 5  │ cancelled │
        │       └── SendCavcRemandProcessedLetterTask │ 6  │ on_hold   │
        │           └── CavcPoaClarificationTask      │ 12 │ assigned  │
        ├── TranslationTask                           │ 7  │ assigned  │
        ├── TranscriptionTask                         │ 8  │ assigned  │
        ├── PrivacyActTask                            │ 9  │ assigned  │
        └── IhpColocatedTask                          │ 10 │ on_hold   │
            └── IhpColocatedTask                      │ 11 │ assigned  │
                                                      └────────────────┘
```

File: spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb [hunk 8]
```diff
@@ -119,6 +119,10 @@
         SEND_IHP_TO_COLOCATED_BLOCKING_DISTRIBUTION: [IhpColocatedTask, Colocated]
       }
 
+      blocking_admin_actions = {
+        CLARIFY_POA_BLOCKING_CAVC: [CavcPoaClarificationTask, CavcLitigationSupport]
+      }
+
```
Comment (yoomlam) [inline spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:125]: I would appreciate code comments explaining the purpose for initializing `completed_distribution_task` on lines 109-111.

File: spec/feature/queue/cavc_task_queue_spec.rb (hunk 7 above)
Comment (hschallhorn) [reply spec/feature/queue/cavc_task_queue_spec.rb:94]: the user story states that all other admin actions should block distribution and poa clarification should block the letter

File: spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb (hunk 8 above)
Comment (hschallhorn) [reply spec/models/tasks/send_cavc_remand_processed_letter_task_spec.rb:125]: Will add in the OG PR!

File: spec/feature/queue/cavc_task_queue_spec.rb (hunk 7 above)
Comment (hschallhorn) [reply spec/feature/queue/cavc_task_queue_spec.rb:94]: #15286 contains the refined ticket and changes to the implementation

**Review by yoomlam — APPROVED** [review]
🎉

File: app/models/task.rb [hunk 9]
```diff
@@ -171,6 +170,12 @@ def create_from_params(params, user)
       child
     end
 
+    def parent_of_same_type_has_same_assignee(parent_task, params)
+      parent_task.assigned_to_id == params[:assigned_to_id] &&
+        parent_task.assigned_to_type == params[:assigned_to_type] &&
+        parent_task.type == params[:type]
+    end
+
```
Comment (hschallhorn) [inline app/models/task.rb:178]: Fix a bit of this validation. There are plenty of places where we expect the assignee of a child task to be the same as the parent. We add a check here for the type as I believe that's the true intent of this validation. 

This allows cavc admins to create tasks assigned to their org as children for other tasks assigned to their org. (see below)

```ruby
                                                  ┌────────────────────────────────────────────────────────────────────────────────────────────┐
Appeal 932 (evidence_submission) ──────────────── │ ID   │ STATUS   │ ASGN_BY                │ ASGN_TO               │ UPDATED_AT              │
└── RootTask                                      │ 2823 │ on_hold  │                        │ Bva                   │ 2020-11-23 21:27:48 UTC │
    └── DistributionTask                          │ 2824 │ on_hold  │                        │ Bva                   │ 2020-11-23 21:27:48 UTC │
        └── CavcTask                              │ 2825 │ on_hold  │                        │ Bva                   │ 2020-11-23 21:27:48 UTC │
            └── SendCavcRemandProcessedLetterTask │ 2826 │ on_hold  │                        │ CavcLitigationSupport │ 2020-11-25 15:54:49 UTC │
                └── CavcPoaClarificationTask      │ 3379 │ assigned │ CAVC_LIT_SUPPORT_ADMIN │ CavcLitigationSupport │ 2020-11-25 15:54:48 UTC │
                                                  └────────────────────────────────────────────────────────────────────────────────────────────┘
```

File: app/models/task.rb [hunk 10]
```diff
@@ -160,8 +160,7 @@ def any_recently_updated(*tasks_arrays)
 
     def create_from_params(params, user)
       parent_task = Task.find(params[:parent_id])
-      fail Caseflow::Error::ChildTaskAssignedToSameUser if parent_task.assigned_to_id == params[:assigned_to_id] &&
-                                                           parent_task.assigned_to_type == params[:assigned_to_type]
+      fail Caseflow::Error::ChildTaskAssignedToSameUser if parent_of_same_type_has_same_assignee(parent_task, params)
```
Comment (yoomlam) [inline app/models/task.rb:163]: Update error message?
https://github.com/department-of-veterans-affairs/caseflow/blob/689300c737ae8799f23c4731d09a378db5321cea/lib/caseflow/error.rb#L258
to:
```ruby
      @message = "A task of the same type cannot be assigned to the same user as the parent."
```

**Comment by yoomlam** [comment]
Fewer clicks to get a simpler task tree -- Sweet! 
```ruby
                                                  ┌─────────────────────────────────────────┐
Appeal 918 (evidence_submission) ──────────────── │ ID   │ STATUS   │ ASGN_TO               │
└── RootTask                                      │ 2743 │ on_hold  │ Bva                   │
    └── DistributionTask                          │ 2744 │ on_hold  │ Bva                   │
        └── CavcTask                              │ 2745 │ on_hold  │ Bva                   │
            └── SendCavcRemandProcessedLetterTask │ 2746 │ on_hold  │ CavcLitigationSupport │
                └── CavcPoaClarificationTask      │ 3378 │ assigned │ CavcLitigationSupport │
                                                  └─────────────────────────────────────────┘
```
However, the case shows up in both the Unassigned and Assigned tabs of the Team view.

**Comment by yoomlam** [comment]
Should Regina (CAVC_LIT_SUPPORT_USER1, who is a non-admin) have these actions after adding CavcPoaClarificationTask?
image
```ruby
                                                      ┌──────────────────────────────────────────┐
Appeal 920 (ES 201124-919) ────────────────────────── │ ID   │ STATUS   │ ASGN_TO                │
└── RootTask                                          │ 2755 │ on_hold  │ Bva                    │
    └── DistributionTask                              │ 2756 │ on_hold  │ Bva                    │
        └── CavcTask                                  │ 2757 │ on_hold  │ Bva                    │
            └── SendCavcRemandProcessedLetterTask     │ 2758 │ on_hold  │ CavcLitigationSupport  │
                └── SendCavcRemandProcessedLetterTask │ 3379 │ on_hold  │ CAVC_LIT_SUPPORT_USER1 │
                    └── CavcPoaClarificationTask      │ 3380 │ assigned │ CavcLitigationSupport  │
                                                      └──────────────────────────────────────────┘
```

**Review by yoomlam — APPROVED** [review]
🎉

**Pull request merged**